extensions/webchat/src/node/directline/middleware/transcriptHandler.ts

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { StatusCodes } from 'http-status-codes';
+import * as express from 'express';
+
+import { DLServerState } from '../store/dlServerState';
+import { Conversation } from '../store/entities/conversation';
+import { textItem } from '../utils/helpers';
+import { writeFile, mkdirp } from '../utils/fileOperations';
+
+export const saveTranscriptHandler = (state: DLServerState) => {
+  return async (req: express.Request, res: express.Response): Promise<void> => {
+    const { fileSavePath } = req.body;
+    const conversation: Conversation = (req as any).conversation;
+    if (!conversation) {
+      res.status(StatusCodes.NOT_FOUND).send('Conversation not found');
+      const logItem = textItem('Error', 'Cannot find a matching conversation.');
+      state.dispatchers.logToDocument(req.params.conversationId, logItem);
+      return;
+    }
+
+    try {
+      if (!fileSavePath?.length) {
+        res.status(StatusCodes.BAD_REQUEST).send('Invalid file path to save the transcript.');
+      }
+      await mkdirp(fileSavePath);
+      const transcripts = await conversation.getTranscript();
+      const contentsToWrite = typeof transcripts === 'object' ? JSON.stringify(transcripts, null, 2) : transcripts;
+      writeFile(fileSavePath, contentsToWrite);
+      res.status(StatusCodes.CREATED).json({
+        path: req.query.fileSavePath,
+        message: 'Transcript has been saved to disk sunccesfully',
+      });
+    } catch (ex) {
+      res.status(StatusCodes.INTERNAL_SERVER_ERROR).send(ex);
+    }
+  };
+};
+
+export const getTranscriptHandler = (state: DLServerState) => {
```

Comment:
do we need any of this? Do / will we have UI to support saving transcripts?

Reply:
Yep. we do have UI to save the conversation as a transcript for R12 @tonyanziano

Reply:
![](https://user-images.githubusercontent.com/13004779/107718448-cd43ee80-6c8a-11eb-8c60-c58ff1c4b14a.png)